src/emit.cpp

```diff
@@ -0,0 +1,76 @@
+#include "emit.h"
+
+#include <map>
+
+#include "debug.h"
+#include "generic_factory.h"
+
+template <>
+const emit_id string_id<emit>::NULL_ID( "null" );
+
+static std::map<emit_id, emit> emits_all;
+
+template<>
+bool string_id<emit>::is_valid() const
+{
+    const auto found = emits_all.find( *this );
+    if( found == emits_all.end() ) {
+        return false;
+    }
+    return found->second.field() != fd_null;
+}
+
+template<>
+const emit &string_id<emit>::obj() const
+{
+    const auto found = emits_all.find( *this );
+    if( found == emits_all.end() ) {
+        debugmsg( "Tried to get invalid emission data: %s", c_str() );
```

Comment:
It shouldn't. Those `emit` objects that emit `fd_null` are actually valid, they simply emit nothing. `map::propagate_field()`, in turn, should have a guard clause to prevent propagating fields of nothing.

Also, it's better to invoke the validity check once from `emit::check_consistency()` and never again. The function clearly belongs to `emit` as non-static member (`e.[method]` all over the function signal that). `generic_factory` can do many useful things apart from sparing boilerplate and invoking `whatever::check()`: cache to mitigate map lookup, track and report duplicates, edit/overwrite items controllably [e.g. by mods] e.t.c. All for free.